helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -840,11 +808,31 @@ public boolean completeSwapIfPossible(String clusterName, String instanceName,
       return false;
     }
 
-    // Complete the swap by removing the InstanceOperation for the SWAP_IN node and disabling the SWAP_OUT node.
-    setInstanceOperation(clusterName, swapInInstanceConfig.getInstanceName(), null);
-    enableInstance(clusterName, swapOutInstanceConfig.getInstanceName(), false);
+    BaseDataAccessor<ZNRecord> baseAccessor = new ZkBaseDataAccessor<>(_zkClient);
+    String swapInInstanceConfigPath =
+        PropertyPathBuilder.instanceConfig(clusterName, swapInInstanceConfig.getInstanceName());
+    String swapOutInstanceConfigPath =
+        PropertyPathBuilder.instanceConfig(clusterName, swapOutInstanceConfig.getInstanceName());
+
+    return baseAccessor.multiSet(Map.of(swapInInstanceConfigPath, currentData -> {
+      if (currentData == null) {
+        throw new HelixException("Cluster: " + clusterName + ", instance: " + instanceName
+            + ", SWAP_IN instance config is null");
+      }
+
+      InstanceConfig config = new InstanceConfig(currentData);
+      config.setInstanceOperation(InstanceConstants.InstanceOperation.ENABLE);
```

Comment:
This should copy the whole instance config to the new node, expose a flag for whether to include user fields or not.

Exclude HELIX_HOST.